components/Init/src/virtio_con.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2022, UNSW (ABN 57 195 873 179)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <autoconf.h>
+
+#include <utils/util.h>
+
+#include <camkes.h>
+#include <camkes/dataport.h>
+
+#include <sel4vmmplatsupport/drivers/virtio_con.h>
+#include <sel4vmmplatsupport/device.h>
+
+#include <sel4vm/guest_irq_controller.h>
+#include <sel4vm/boot.h>
+
+#include <platsupport/serial.h>
+
+#include <virtqueue.h>
+#include <camkes/virtqueue.h>
+
+#include "virtio_con.h"
+#include "virtio_irq.h"
+
+static virtio_con_t *virtio_con;
+
+typedef struct virtio_con_cookie {
+    virtio_con_t *virtio_con;
+    vm_t *vm;
+} virtio_con_cookie_t;
+
+/* Configured in camkes */
+#define NUM_PORTS ARRAY_SIZE(serial_layout)
+
+/* 4088 because the serial_shmem_t struct has to be 0x1000 bytes big */
+#define BUFSIZE 4088
+
+/* This struct occupies exactly 1 page and represents the data in the shmem region */
+typedef struct serial_shmem {
+    uint32_t head;
+    uint32_t tail;
+    char buf[BUFSIZE];
+} serial_shmem_t;
+compile_time_assert(serial_shmem_1k_size, sizeof(serial_shmem_t) == 0x1000);
+
+/*
+ * Represents the virtqueues used in a given link between
+ * two VMs.
+ */
+typedef struct serial_conn {
+    void (*notify)(void);
+    volatile serial_shmem_t *recv_buf;
+    volatile serial_shmem_t *send_buf;
+} serial_conn_t;
+static serial_conn_t connections[NUM_PORTS];
+
+/* This is a temporary buffer used to stage data before completing the virtio RX */
+char temp_rx_buf[BUFSIZE];
+
+void virtio_console_putchar(int port, virtio_emul_t *con, char *buf, int len) WEAK;
+static int virtio_con_notify_recv(int port)
+{
+    volatile serial_shmem_t *buffer = connections[port].recv_buf;
+    int count = 0;
+    while (buffer->head != buffer->tail) {
+        temp_rx_buf[count] = buffer->buf[buffer->head];
+        buffer->head = (buffer->head + 1) % BUFSIZE;
+        count++;
+    }
+
+    assert(count <= BUFSIZE);
+    virtio_console_putchar(port, virtio_con->emul, temp_rx_buf, count);
+}
+
+void handle_serial_console(vm_t *vmm)
+{
+    /* Poll each ring buffer to see if data was added */
+    for (int i = 0; i < NUM_PORTS; i++) {
+        if (connections[i].recv_buf->head != connections[i].recv_buf->tail) {
+            virtio_con_notify_recv(i);
+        }
+    }
+}
+
+static void tx_virtcon_forward(int port, char c)
+{
+#if CONFIG_NUM_DOMAINS > 1 && CONFIG_PRINTING
+    seL4_DebugPutChar(c);
+#else
+    if (port >= NUM_PORTS) {
+        return;
+    }
+
+    volatile serial_shmem_t *batch_buf = connections[port].send_buf;
```

Comment:
Don't think this needs to be volatile.

Reply:
I think it must be volatile because this is shared memory. But I wonder if we need even stronger access constraints then with barriers, to really ensure the proper visibility. It might be worth moving the whole shares mem access to a dedicated file then that wraps all these details behind a nice API, that allows re-using it then also.

Reply:
It's shared memory, yes, but it doesn't need 'volatile' if the memory barriers are correct, and 'volatile' may hurt performance.